src/main/java/org/apache/commons/collections4/iterators/ZippedTupleIterator.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections4.iterators;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import org.apache.commons.collections4.iterators.ZippedTupleIterator.ZippedTuple;
+
+/**
+ * Provides a iteration over the elements contained in a pair of Iterators.
+ *
+ * <p>
+ * Given two {@link Iterator} instances {@code A} and {@code B}, the {@link #next} method on this
+ * iterator provide a Pair of {@code A.next()} and {@code B.next()} until one of the iterators is
+ * exhausted.
+ * </p>
+ * Example usage:
+ * <pre>{@code
+ *   List<Integer> studentIds = ...
+ *   List<String> studentNames = ...
+ *
+ *   ZippedTupleIterator<ZippedTuple<Integer, String>> zippedIterator =
+ *     ZippedTupleIterator.ofIterables(studentIds, studentNames);
+ *
+ *   while (zippedIterator.hasNext()) {
+ *     ZippedTuple<Integer, String> item = zippedIterator.next();
+ *     ...
+ *   }
+ * }</pre>
+ *
+ * @param <L> the left elements' type
+ * @param <R> the right elements' type
+ */
+public class ZippedTupleIterator<L, R> implements Iterator<ZippedTuple<L, R>> {
```

Comment:
Why is this prefixed with "zip" when it has nothing to do with zip files?

Reply:
I think it is related to [zipper](https://en.wikipedia.org/wiki/Zipper_(data_structure)). Like Python stdlib's [zip function](https://docs.python.org/3/library/functions.html#zip). Scala and Clojure also have a similar function I think.

Reply:
Maybe but the prefix here is "zipped" not "zipper", terrible name IMO.

Reply:
I'm not good naming things, so can't really suggest a better one. The code example in the Python docs uses `zipped` too for the variable that contains the tuples returned with `zip` (that were zipped), so that should sound OK for users coming from Python (or JS using Lodash's [zip/unzip](https://lodash.com/docs/#unzip) which I'm using in my current project). But if that's confusing for other users, we can try to think in a better name I think.

Reply:
Updated to use `PairedIterator` and `PairedIterable` names. Hope is acceptable to all.